compiler/src/iree/compiler/Codegen/Common/GPU/GPUTileSwizzleUtils.cpp

```diff
@@ -0,0 +1,129 @@
+
+#include "iree/compiler/Codegen/Common/GPU/GPUTileSwizzleUtils.h"
+#include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.h"
+
+namespace mlir::iree_compiler {
+
+// Given an `expandShape` vector-of-vectors describing the mapping from source
+// dimensions to expanded dimensions, returns the index of the first expanded
+// dimension corresponding to the given source dimension index.
+static int64_t
+getExpandedDimFirstIdx(const SmallVector<SmallVector<int64_t>> &expandShape,
+                       int64_t srcIndex) {
+  int dstIndexFirst = 0;
+  for (int i = 0; i < srcIndex; ++i) {
+    dstIndexFirst += expandShape[i].size();
+  }
+  return dstIndexFirst;
+}
+
+void unroll(TileSwizzle &swizzle, int srcIndex, int unrollFactor) {
+  assert(unrollFactor > 1);
+  int dstIndexFirst = getExpandedDimFirstIdx(swizzle.expandShape, srcIndex);
+
+  // The new unrolling dimension is inserted at the start of the expandShape
+  // dimensions group corresponding to srcIndex.
+  swizzle.expandShape[srcIndex].insert(swizzle.expandShape[srcIndex].begin(),
+                                       unrollFactor);
+  // Since we are not interleaving here, generating side-by-side copies of the
+  // original layout, the new unrolling dimension is the new outermost
+  // dimension. Existing entries get shifted to make room for it.
+  for (auto &p : swizzle.permutation) {
+    p += (p >= dstIndexFirst);
+  }
+  swizzle.permutation.insert(swizzle.permutation.begin(), dstIndexFirst);
+}
+
+void interleave(TileSwizzle &swizzle, int srcIndex,
+                int expandedDimIndexToInterleaveAt) {
+  // Compute which inner dimension to permute the current outer dimension into.
+  int dstIndexFirst = getExpandedDimFirstIdx(swizzle.expandShape, srcIndex);
+  int dstIndexToInterleaveAt = dstIndexFirst + expandedDimIndexToInterleaveAt;
+
+  SmallVector<int64_t> outPermutation(swizzle.permutation.size());
+  // The leading dimension, permutation[0], gets moved inwards to the
+  // position that we just computed, dstIndexToInterleaveAt.
+  outPermutation[dstIndexToInterleaveAt] = swizzle.permutation[0];
+  // Outer dimensions get shifted outwards to fill the gap.
+  for (int i = 0; i < dstIndexToInterleaveAt; ++i) {
+    outPermutation[i] = swizzle.permutation[i + 1];
+  }
+  // Inner dimensions don't change. That is to say that we only interleave
+  // at `targetInterleavedElements` granularity, we don't swizzle further
+  // internally to that.
+  for (int i = dstIndexToInterleaveAt + 1; i < outPermutation.size(); ++i) {
+    outPermutation[i] = swizzle.permutation[i];
+  }
+  swizzle.permutation = outPermutation;
+}
+
+// Returns the permutation of indices that sorts `v` with the given comparator.
+template <template <typename U> class Comparator, typename T>
+static SmallVector<int64_t> getSortingPermutation(ArrayRef<T> v) {
+  using P = std::pair<int64_t, T>;
+  SmallVector<P> pairs;
+  for (auto x : v) {
+    pairs.push_back({pairs.size(), x});
+  }
```

Comment:
nice!!

Reply:
IMO, we don't need the reserve because SmallVector should already allocate a ~64 bytes buffer for you. It is good when you're using `std::vector`. It is unnecessary for most of cases in `SmallVector`, IMO.